src/main/java/gov/usgs/util/CryptoUtils.java

```diff
@@ -195,37 +175,138 @@ public static Cipher getDecryptCipher(final Key key)
+				signature = Signature.getInstance(DSA_SIGNATURE_ALGORITHM);
+			} else if (key instanceof RSAPrivateKey || key instanceof RSAPublicKey) {
+				signature = Signature.getInstance(RSA_SIGNATURE_ALGORITHM);
+			}
+		} else if (version == Version.SIGNATURE_V2) {
+			if (key instanceof DSAPrivateKey || key instanceof DSAPublicKey) {
+				signature = Signature.getInstance(SIGNATURE_V2_DSA_ALGORITHM);
+			} else if (key instanceof RSAPrivateKey || key instanceof RSAPublicKey) {
+				signature = Signature.getInstance(SIGNATURE_V2_RSA_ALGORITHM);
+			}
 		} else {
-			throw new InvalidKeyException("Expected a DSA or RSA key.");
+			throw new IllegalArgumentException("Unexpected signature version " + version);
+		}
+		return signature;
+	}
+
+	public static void configureSignature(final Key key, final Version version,
+			final Signature signature) throws InvalidAlgorithmParameterException {
+		if (version == Version.SIGNATURE_V2
+				&& (key instanceof RSAPrivateKey || key instanceof RSAPublicKey)) {
+			int keySize;
+			if (key instanceof RSAPrivateKey) {
+				keySize = ((RSAPrivateKey)key).getModulus().bitLength();
+			} else {
+				keySize = ((RSAPublicKey)key).getModulus().bitLength();
+			}
+			// match python cryptography calculation:
+			// https://github.com/pyca/cryptography/blob/b16561670320c65a18cce41d0db0c42ab68350a9/src/cryptography/hazmat/primitives/asymmetric/padding.py#L73
+			// 32 = (sha)256 / 8
+			int maxSaltLength = ((keySize + 6) / 8) - 32 - 2;
+			signature.setParameter(
+					new PSSParameterSpec("SHA-256", "MGF1",
+							MGF1ParameterSpec.SHA256, maxSaltLength, 1));
 		}
 	}
 
+	/**
+	 * A convenience method that chooses a signature algorithm based on the key
+	 * type. Works with DSA and RSA keys.
+	 * @throws InvalidAlgorithmParameterException
+	 */
+	public static String sign(final PrivateKey privateKey, final byte[] data)
+			throws InvalidAlgorithmParameterException, InvalidKeyException,
+			NoSuchAlgorithmException, SignatureException {
+		// use v1 by default
+		return sign(privateKey, data, Version.SIGNATURE_V1);
+	}
+
+	/**
+	 * Generate a signature.
+	 *
+	 * @param algorithm
+	 *            signature algorithm to use, as accepted by
+	 *            Signature.getInstance.
+	 * @param privateKey
+	 *            private key to use, should be acceptable by signature
+	 *            instance.
+	 * @param data
+	 *            data/hash to sign.
+	 * @param version
+	 *            the signature version.
+	 * @return signature as hex encoded string.
+	 * @throws NoSuchAlgorithmException
+	 * @throws InvalidKeyException
+	 * @throws SignatureException
+	 */
+	public static String sign(final PrivateKey privateKey, final byte[] data,
+			final Version version) throws InvalidAlgorithmParameterException,
+			InvalidKeyException, NoSuchAlgorithmException, SignatureException {
+		final Signature signature = getSignature(privateKey, version);
+		signature.initSign(privateKey);
```

Comment:
... or maybe check the signature here is not null ...

Reply:
Added check/exception in getSignature method.